crates/criticalup-core/src/download_server_client.rs

```diff
@@ -126,42 +126,48 @@ impl DownloadServerClient {
         let mut current_retries: u8 = 0;
         let mut current_retry_backoff = CLIENT_RETRY_BACKOFF;
 
-        // This `try_clone()` will be `None` if request body is a stream.
-        let req_outer = req.try_clone().ok_or(Error::RequestCloningFailed)?;
-        let mut response_result = self.client.execute(req_outer);
+        while current_retries < CLIENT_MAX_RETRIES {
+            // This `try_clone()` will be `None` if request body is a stream.
+            let response_result = self
+                .client
+                .execute(req.try_clone().ok_or(Error::RequestCloningFailed)?);
 
-        while current_retries < CLIENT_MAX_RETRIES && response_result.is_err() {
-            let req = req.try_clone().ok_or(Error::RequestCloningFailed)?;
-
-            response_result = self.client.execute(req);
             if response_result.is_ok() {
-                break;
-            }
+                let response = response_result?;
 
+                if response.status() == StatusCode::OK {
+                    return Ok(response);
+                } else {
+                    return Err(self.err_from_response_with_error_code(&response));
+                }
+            }
             current_retries += 1;
             current_retry_backoff *= 2;
             thread::sleep(Duration::from_millis(current_retry_backoff));
         }
 
-        let response = response_result.map_err(|e| Error::DownloadServerError {
-            kind: DownloadServerError::Network(e),
-            url,
-        })?;
-
-        Err(self.err_from_response(
-            &response,
-            match response.status() {
-                StatusCode::OK => return Ok(response),
+        self.client
```

Comment:
This is kind of weird. If we've already retried the defined number of times, we should return an error suggesting retries exceeded, with context from the last request's error.

I believe you added this to get a more rich error, but I don't really understand why we can't use the `response_result` from above (L136) instead of forcing a new request.